<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="author" content="Brent Pedersen" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.6/css/selectize.bootstrap3.min.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/css/bootstrap.min.css">
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.6/js/standalone/selectize.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.bundle.min.js"></script>
    <style>
    html,body { height: 100%; margin: 0px; padding: 0px; }
    .form-control.selectize-control {
        padding: 5px 6px 0px;
        height: unset !important;
    }
    .remove-single {
        color: gray !important;
        top: -1px !important;
        font-size: 20px !important;
    }
    </style>
</head>
<body>

<nav class="navbar navbar-expand-lg navbar-dark p-0" style="background-color:#4b636e !important">
  <a class="navbar-brand ml-2" href="https://github.com/brentp/somalier">somalier</a>
  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#toggler" aria-controls="toggler" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  <div class="collapse navbar-collapse" id="toggler">
    <div class="navbar-nav mr-auto">
    </div>
    <div>
        <div class="input-group" style="width:300px">
            <input type="text" class="form-control" id="sample-search" aria-label="sample-search">
        </div>
    </div>
  </div>
</nav>
<div class="container-fluid h-100">
    <div class="row h-75">
        <div class="col-6 h-100 bg-light pt-2 border-right">
            <div class="row">
                <div class="col">
                    <h5>Sample to Sample Relatedness</h5>
                </div>
            </div>
            <div class="row pb-2">
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">X Axis</div>
                        </div>
                        <select class="custom-select" id="plotax_select">
                            <option value="shared_hets" title="the number of sites where both samples are heterozygotes">Shared hets</option>
                            <option value="shared_hom_alts" title="the number of sites where both samples are homozygous alternate">Shared hom-alts</option>
                            <option value="concordance" title="a relatedness estimate that is not affected by loss-of-heterozygosity">Homozygous concordance</option>
                            <option value="relatedness" title="relatedness of a pair of samples (2*kinship coefficient")>Relatedness</option>
                            <option value="ibs0" title="the number of sites where one sample is hom-ref and another is hom-alt" selected>IBS0</option>
                            <option value="ibs2" title="the number of sites where the samples have the same genotype">IBS2</option>
                        </select>
                    </div>
                </div>
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">Y Axis</div>
                        </div>
                        <select class="custom-select" id="plotay_select">
                            <option value="shared_hets" title="the number of sites where both samples are heterozygotes">Shared hets</option>
                            <option value="shared_hom_alts" title="the number of sites where both samples are homozygous alternate">Shared hom-alts</option>
                            <option value="concordance" title="a relatedness estimate that is not affected by loss-of-heterozygosity">Homozygous concordance</option>
                            <option value="relatedness" title="relatedness of a pair of samples (2*kinship coefficient")>Relatedness</option>
                            <option value="ibs0" title="the number of sites where one sample is hom-ref and another is hom-alt">IBS0</option>
                            <option value="ibs2" title="the number of sites where the samples have the same genotype" selected>IBS2</option>
                        </select>
                    </div>
                </div>
            </div>
            <div class="row h-100">
                <div class="col h-100">
                    <div id="plota" style="height: 90%"></div>
                </div>
            </div>
        </div>

        <div class="col-6 h-100 bg-light pt-2">
            <div class="row pt-0 pb-0">
                <div class="col pt-0 pb-0">
                    <h5>Sample Depth Metrics</h5>
                </div>
            </div>
            <div class="row pb-2">
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">X Axis</div>
                        </div>
                        <select class="custom-select" id="plotbx_select">
                            <option value="gt_depth_mean">Mean depth of genotyped sites</option>
                            <option value="depth_mean">Mean depth of all sites</option>
                            <option value="ab_mean">Mean allele balance</option>
                            <option value="pct_other_alleles">% reads with neither REF nor ALT</option>
                            <option value="n_hom_ref">Number of 0/0 sites</option>
                            <option value="n_het">Number of 0/1 sites</option>
                            <option value="n_hom_alt">Number of 1/1 sites</option>
                            <option value="n_unknown">Number of unknown sites</option>
                            <option value="n_known">Number of known sites</option>
                            <option value="p_middling_ab">Proportion sites with AB &lt; 0.1 or AB &gt; 0.9</option>
                            <option value="x_depth_mean">Scaled mean depth on X</option>
                            <option value="x_hom_ref">Number of 0/0 sites on X</option>
                            <option value="x_het" selected >Number of 0/1 sites on X</option>
                            <option value="x_hom_alt">Number of 1/1 sites on X</option>
                            <option value="y_depth_mean">Scaled mean depth on Y</option>
                        </select>
                    </div>
                </div>
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">Y Axis</div>
                        </div>
                        <select class="custom-select" id="plotby_select">
                            <option value="gt_depth_mean">Mean depth of genotyped sites</option>
                            <option value="depth_mean">Mean depth of all sites</option>
                            <option value="ab_mean">Mean allele balance</option>
                            <option value="pct_other_alleles">% reads with neither REF nor ALT</option>
                            <option value="n_hom_ref">Number of 0/0 sites</option>
                            <option value="n_het">Number of 0/1 sites</option>
                            <option value="n_hom_alt">Number of 1/1 sites</option>
                            <option value="n_unknown">Number of unknown sites</option>
                            <option value="n_known">Number of known sites</option>
                            <option value="p_middling_ab">Proportion sites with AB &lt; 0.1 or AB &gt; 0.9</option>
                            <option value="x_depth_mean" selected>Scaled mean depth on X</option>
                            <option value="x_hom_ref">Number of 0/0 sites on X</option>
                            <option value="x_het">Number of 0/1 sites on X</option>
                            <option value="x_hom_alt">Number of 1/1 sites on X</option>
                            <option value="y_depth_mean">Scaled mean depth on Y</option>
                        </select>
                    </div>
                </div>
            </div>
            <div class="row h-100">
                <div class="col h-100">
                    <div id="plotb" style="height: 90%"></div>
                </div>
            </div>
        </div>
    </div>
    <div class="row h-25">
        <div class="col-12 h-100 bg-light pt-2 border-right">
            Choose pre-set selection for sample plot:
            <ul>
            <li><a href="javascript:set_plotb('x_het', 'x_depth_mean')">Sex QC: Number of 0/1 sites on X vs scaled mean depth on chrX</a></li>

            <li><a href="javascript:set_plotb('x_depth_mean', 'y_depth_mean')">Sex ploidy QC: Scaled mean depth on chrX vs. chrY</a></li>
            <li><a href="javascript:set_plotb('n_unknown', 'p_middling_ab')">General QC: number of sites with unknown genotype vs. proportion of hets with allele balance outside of 0.1 - 0.9</a></li>
            <li><a href="javascript:set_plotb('gt_depth_mean', 'ab_mean')">Depth and allele QC: mean depth vs. mean het allele-balance</a></li>
            </ul>
        </div>
    </div>
</div>

<script>
var nan = NaN; // hack to support json dumped with NaN values.
var sample_data = [{"sample":"U084466CM2014D19622","sex":"unknown","gt_depth_mean":43.40475335249054,"depth_mean":41.70720202485057,"ab_mean":0.5279733984166023,"pct_other_alleles":0.0,"n_hom_ref":4456,"n_het":6729,"n_hom_alt":5376,"n_unknown":823,"n_known":16561,"p_middling_ab":0.001550743782666131,"x_depth_mean":1.024347220694143,"x_hom_ref":147,"x_het":1,"x_hom_alt":177,"y_depth_mean":0.9906749072111173},
{"sample":"U084466PF2014D19620","sex":"unknown","gt_depth_mean":42.02975335249042,"depth_mean":40.38569949378737,"ab_mean":0.5260103988031531,"pct_other_alleles":0.0,"n_hom_ref":4517,"n_het":6673,"n_hom_alt":5375,"n_unknown":819,"n_known":16565,"p_middling_ab":0.001952003674359858,"x_depth_mean":1.968476169901964,"x_hom_ref":98,"x_het":119,"x_hom_alt":107,"y_depth_mean":0.0},
{"sample":"U084466PM2014D19812","sex":"unknown","gt_depth_mean":41.0175396587846,"depth_mean":39.41543948458372,"ab_mean":0.5296667364337306,"pct_other_alleles":0.0,"n_hom_ref":4428,"n_het":6758,"n_hom_alt":5386,"n_unknown":812,"n_known":16572,"p_middling_ab":0.00126393197747903,"x_depth_mean":1.048482507660492,"x_hom_ref":159,"x_het":0,"x_hom_alt":165,"y_depth_mean":0.9101797339357913},
{"sample":"U143920CM2016D22827","sex":"unknown","gt_depth_mean":42.75184098664901,"depth_mean":41.07708237459715,"ab_mean":0.5288930884285896,"pct_other_alleles":0.0,"n_hom_ref":4409,"n_het":6799,"n_hom_alt":5367,"n_unknown":809,"n_known":16575,"p_middling_ab":0.001493394600804135,"x_depth_mean":1.067025019223456,"x_hom_ref":144,"x_het":0,"x_hom_alt":180,"y_depth_mean":1.013601574417951},
{"sample":"U143920PF2016D22831","sex":"unknown","gt_depth_mean":40.15953307392967,"depth_mean":38.59094569719307,"ab_mean":0.5250359521904197,"pct_other_alleles":0.0,"n_hom_ref":4411,"n_het":6921,"n_hom_alt":5243,"n_unknown":809,"n_known":16575,"p_middling_ab":0.001321307519963233,"x_depth_mean":1.98370324990177,"x_hom_ref":81,"x_het":143,"x_hom_alt":98,"y_depth_mean":0.0},
{"sample":"U143920PF2016D24514","sex":"unknown","gt_depth_mean":40.61167315175116,"depth_mean":39.02542567878518,"ab_mean":0.5277470400273383,"pct_other_alleles":0.0,"n_hom_ref":4363,"n_het":6907,"n_hom_alt":5286,"n_unknown":828,"n_known":16556,"p_middling_ab":0.001722751808889399,"x_depth_mean":1.933597896215687,"x_hom_ref":78,"x_het":136,"x_hom_alt":105,"y_depth_mean":0.0},
{"sample":"U192125CF2019D01182","sex":"unknown","gt_depth_mean":40.13008081412753,"depth_mean":38.56264381040054,"ab_mean":0.5246112506175251,"pct_other_alleles":0.0,"n_hom_ref":4536,"n_het":6684,"n_hom_alt":5352,"n_unknown":812,"n_known":16572,"p_middling_ab":0.001321307519963233,"x_depth_mean":1.959770985494545,"x_hom_ref":88,"x_het":139,"x_hom_alt":92,"y_depth_mean":0.0},
{"sample":"U192125PF2019D11568","sex":"unknown","gt_depth_mean":40.57210416043105,"depth_mean":38.98740220892797,"ab_mean":0.5220298542937063,"pct_other_alleles":0.0,"n_hom_ref":4509,"n_het":6817,"n_hom_alt":5250,"n_unknown":808,"n_known":16576,"p_middling_ab":0.001378676470588235,"x_depth_mean":1.933491546882669,"x_hom_ref":93,"x_het":125,"x_hom_alt":105,"y_depth_mean":0.0},
{"sample":"U192125PM2019D11570","sex":"unknown","gt_depth_mean":40.89157636352753,"depth_mean":39.28997929130244,"ab_mean":0.5275747228186761,"pct_other_alleles":0.0,"n_hom_ref":4501,"n_het":6653,"n_hom_alt":5418,"n_unknown":812,"n_known":16572,"p_middling_ab":0.001952003674359858,"x_depth_mean":1.008481300067,"x_hom_ref":159,"x_het":0,"x_hom_alt":164,"y_depth_mean":0.880376918707138}]
var input = [{"expected_relatedness":0.0,"text":["U084466CM2014D19622<br>U084466PF2014D19620","U084466CM2014D19622<br>U084466PM2014D19812","U084466CM2014D19622<br>U143920CM2016D22827","U084466CM2014D19622<br>U143920PF2016D22831","U084466CM2014D19622<br>U143920PF2016D24514","U084466CM2014D19622<br>U192125CF2019D01182","U084466CM2014D19622<br>U192125PF2019D11568","U084466CM2014D19622<br>U192125PM2019D11570","U084466PF2014D19620<br>U084466PM2014D19812","U084466PF2014D19620<br>U143920CM2016D22827","U084466PF2014D19620<br>U143920PF2016D22831","U084466PF2014D19620<br>U143920PF2016D24514","U084466PF2014D19620<br>U192125CF2019D01182","U084466PF2014D19620<br>U192125PF2019D11568","U084466PF2014D19620<br>U192125PM2019D11570","U084466PM2014D19812<br>U143920CM2016D22827","U084466PM2014D19812<br>U143920PF2016D22831","U084466PM2014D19812<br>U143920PF2016D24514","U084466PM2014D19812<br>U192125CF2019D01182","U084466PM2014D19812<br>U192125PF2019D11568","U084466PM2014D19812<br>U192125PM2019D11570","U143920CM2016D22827<br>U143920PF2016D22831","U143920CM2016D22827<br>U143920PF2016D24514","U143920CM2016D22827<br>U192125CF2019D01182","U143920CM2016D22827<br>U192125PF2019D11568","U143920CM2016D22827<br>U192125PM2019D11570","U143920PF2016D22831<br>U143920PF2016D24514","U143920PF2016D22831<br>U192125CF2019D01182","U143920PF2016D22831<br>U192125PF2019D11568","U143920PF2016D22831<br>U192125PM2019D11570","U143920PF2016D24514<br>U192125CF2019D01182","U143920PF2016D24514<br>U192125PF2019D11568","U143920PF2016D24514<br>U192125PM2019D11570","U192125CF2019D01182<br>U192125PF2019D11568","U192125CF2019D01182<br>U192125PM2019D11570","U192125PF2019D11568<br>U192125PM2019D11570"],"ibs0":[1,1,1478,1453,1458,1559,1561,1486,1395,1436,1396,1482,1529,1484,1538,1501,1486,1496,1537,1485,1498,0,881,1487,1474,1477,1,1484,1483,1510,1584,1486,1593,1,0,1499],"ibs2":[9742,9718,7138,7202,7109,7110,7094,7103,7210,7179,7220,7162,7040,6998,7185,7199,7239,7052,7181,6998,7049,9633,8099,7084,7070,7166,9461,7023,7189,7104,7039,7018,7037,9644,9574,7229],"shared_hets":[3297,3317,2802,2877,2837,2769,2825,2709,2740,2765,2827,2843,2691,2709,2745,2841,2922,2833,2806,2756,2706,3396,3072,2753,2797,2778,3371,2780,2926,2821,2837,2847,2818,3287,3180,2822],"shared_hom_alts":[3591,3583,2448,2439,2444,2427,2365,2503,2536,2491,2444,2449,2431,2381,2510,2511,2454,2395,2469,2404,2503,3494,2859,2450,2423,2480,3398,2393,2413,2405,2382,2360,2413,3475,3567,2471],"concordance":[0.6677209138870239,0.6661086082458496,-0.09465250372886658,-0.08907113969326019,-0.08929247409105301,-0.1291106194257736,-0.1441904753446579,-0.0872395858168602,-0.04725581407546997,-0.07098937779664993,-0.06637421250343323,-0.09742716699838638,-0.1171524673700333,-0.1118095219135284,-0.1053023263812065,-0.09148500114679337,-0.09879839420318604,-0.1129398420453072,-0.1130418553948402,-0.1078095212578773,-0.0915336087346077,0.666412353515625,0.2075293213129044,-0.09790732711553574,-0.1000000014901161,-0.08831749856472015,0.6477207541465759,-0.1096700355410576,-0.1054739654064178,-0.1172992587089539,-0.1486946642398834,-0.1165714263916016,-0.1462353318929672,0.6615238189697266,0.6664798259735107,-0.1003809496760368],"relatedness":[0.4961975812911987,0.4968152940273285,-0.02295081876218319,-0.004284869879484177,-0.01168379839509726,-0.05247331410646439,-0.04422933608293533,-0.0396621935069561,-0.007512583397328854,-0.01602516137063503,0.005194034427404404,-0.01797252148389816,-0.05537115409970284,-0.03872028738260269,-0.05013632401823997,-0.02397974394261837,-0.007373000029474497,-0.0234790313988924,-0.04019196331501007,-0.03176723793148994,-0.04358608275651932,0.4990814924240112,0.1927462667226791,-0.03303191065788269,-0.02237368561327457,-0.02634927816689014,0.4915377795696259,-0.02785597927868366,-0.005870267283171415,-0.0295339860022068,-0.04910979047417641,-0.0183540116995573,-0.05478228628635406,0.4912149608135223,0.4805440008640289,-0.02634336240589619],"n":[16430,16430,16432,16437,16416,16433,16435,16433,16436,16439,16439,16423,16443,16442,16437,16446,16444,16426,16442,16444,16442,16450,16429,16446,16448,16446,16428,16445,16448,16448,16429,16431,16429,16446,16449,16446]}]

// ibs0 and ibs2 are counts. on page-load, we change them to scale by n
    /*
for(var i=0; i< input.length; i++){
    var rel = input[i];
    for(var j=0; j<rel.ibs0.length; j++){
        rel.ibs0[j] /= rel.n[j];
        rel.ibs2[j] /= rel.n[j];
    }
}
*/


var colors = ['rgba(55,126,184,0.7)', 'rgba(228,26,28,0.7)', 'rgba(152,78,163,0.7)', 'rgba(255,127,0,0.7)', 'rgba(166,86,40,0.7)', 'rgba(247,129,191,0.7)', 'rgba(77,175,74,0.7)',]
var size
if (sample_data.length > 700) {
    size = 8
} else if (sample_data.length > 200) {
    size = 9
} else if (sample_data.length > 50) {
    size = 10
} else if (sample_data.length > 20) {
    size = 12
} else {
    size = 15
}

function set_plotb(x, y) {
    jQuery('#plotbx_select').val(x)
    jQuery('#plotby_select').val(y)
    jQuery('#plotby_select, #plotbx_select').trigger('change')
}

function set_xy_data_by_group(input, metric, is_x) {
    let l = {"0.00": "unrelated", "0.50": "parent-child", "0.49": "sib-sib", "1.00": "identical"}
    for(var i = 0; i < input.length; i++) {
        var c = input[i]
        if(c.expected_relatedness != undefined) {
                c.name = l[c.expected_relatedness.toFixed(2)] || c.expected_relatedness.toFixed(3)
        }
        if(is_x) {
            c.x = c[metric]
            c.showlegend = i < input.length - 1 && c.x.length > 0;
        } else {
           c.y = c[metric]
            c.showlegend =i < input.length - 1 &&  c.y.length > 0;
        }
        c.type = 'scattergl'
        c.mode = 'markers'
        if(i < input.length - 1 ){
           c.marker = {size: size, color: colors[i]}
        }

    }
}


var layout_a = {
    autosize: true,
    margin: {t: 30, pad: 0},
    xaxis: {
        title: jQuery("#plotax_select option:selected").text(),
    },
    yaxis: {
        title: jQuery("#plotay_select option:selected").text(),
    },
    hovermode: 'closest',
    showlegend: true,
    legend: {
        xanchor: "right",
        yanchor: "top",
        y: 1,
        x: 1,
        orientation: "h",
        borderwidth: 1,
        bordercolor: '#eeeeee'
    },
}

var layout_b = {
    autosize: true,
    margin: {t: 30, pad: 0},
    xaxis: {
        title: jQuery("#plotbx_select option:selected").text(),
    },
    yaxis: {
        title: jQuery("#plotby_select option:selected").text(),
    },
    hovermode: 'closest',
    showlegend: true,
    legend: {
        xanchor: "right",
        yanchor: "top",
        y: 1,
        x: 1,
        orientation: "h",
        borderwidth: 1,
        bordercolor: '#eeeeee'
    },
}

input.push({mode:"markers", type:'scattergl', hoverinfo:'text+x+y', showlegend:false, name:'hovered', x:[], y:[], hovertext:[], marker:{size:size+3, opacity: 1, color:'#000000'}})
set_xy_data_by_group(input, jQuery('#plotax_select').val(), true)
set_xy_data_by_group(input, jQuery('#plotay_select').val(), false)

// append final trace to hold plot B hover action markers

var traces_b = [
    // gray
    {x:[], y:[], text: [], type: 'scatter', mode: 'markers', marker: {size: size + 3, color: 'rgba(187,187,187,0.9)', line:{color: '#222', width: []}}, showlegend:false, name: "unknown"},
    // light blue
    {x:[], y:[], text: [], type: 'scatter', mode: 'markers', marker: {size: size + 3, color: 'rgba(153,204,255,0.9)', line:{color: '#222', width: []}}, showlegend:true, name: "male"},
    // light red
    {x:[], y:[], text: [], type: 'scatter', mode: 'markers', marker: {size: size + 3, color: 'rgba(255,153,170,0.9)', line:{color: '#222', width: []}}, showlegend:true, name: "female"},
]
var xf = jQuery('#plotbx_select').val()
var yf = jQuery('#plotby_select').val()

for (i in sample_data) {
    var idx = 0
    var sex = sample_data[i].sex
    if (sex == "male") {
        idx = 1
    } else if (sex == "female") {
        idx = 2
    }

    traces_b[idx].x.push(sample_data[i][xf])
    traces_b[idx].y.push(sample_data[i][yf])
    traces_b[idx].text.push("sample:" + sample_data[i].sample)

    // // individual is male; highlight points of low Y coverage
    // if (idx == 1) {
    //     // these samples could be female mislabeled as male
    //     if (sample_data[i]["y_depth_mean"] <= 0.2) {
    //         traces_b[idx].marker.line.width.push(size < 7 ? 1.5 : 2)
    //     } else {
    //         traces_b[idx].marker.line.width.push(0)
    //     }
    // // handle unknown and female
    // } else {
    //     // these samples could be male and mislabeled as female
    //     if (sample_data[i]["y_depth_mean"] > 0.2) {
    //         traces_b[idx].marker.line.width.push(size < 7 ? 1.5 : 2)
    //     } else {
    //         traces_b[idx].marker.line.width.push(0)
    //     }
    // }
}

var pa = document.getElementById("plota")
var pb = document.getElementById("plotb")
Plotly.newPlot(pa, input, layout_a)

pa.on('plotly_hover', function(e) {
    var p = e.points[0]
    var split = p.text.split("<br>", 2)

    Plotly.Fx.hover(pb,
        [
            {curveNumber: 0, pointNumber: unknown_sample_names.indexOf(split[0])},
            {curveNumber: 0, pointNumber: unknown_sample_names.indexOf(split[1])},
            {curveNumber: 1, pointNumber: male_sample_names.indexOf(split[0])},
            {curveNumber: 1, pointNumber: male_sample_names.indexOf(split[1])},
            {curveNumber: 2, pointNumber: female_sample_names.indexOf(split[0])},
            {curveNumber: 2, pointNumber: female_sample_names.indexOf(split[1])},
        ]
    )
})

init_depth_metrics_plot()

pa.on('plotly_unhover', function(e) {
    Plotly.Fx.unhover(pb)
})

function pa_hover(sample) {
    var ht = []
    var xs = []
    var ys = []
    if (sample) {
        var left = sample + "<br>"
        var right = "<br>" + sample
        var otr
        for (var j=0; j < input.length-1; j++) {
            otr = input[j]
 
            for (var k=0; k < otr.x.length; k++) {
                if (otr.text[k].startsWith(left) || otr.text[k].endsWith(right)) {
                    xs.push(otr.x[k])
                    ys.push(otr.y[k])
                    ht.push(otr.text[k])
                }
            }
        }
    }

    input[input.length-1].x = xs
    input[input.length-1].y = ys
    input[input.length-1].hovertext = ht
    input[input.length-1].marker.opacity = 1

    // remove all existing markers so highlights will be drawn on top
    //Plotly.purge(pa)
    // redraw with existing + marker traces
    Plotly.react(pa, input, layout_a)
}

function select_and_hover() {
    // get selection
    var sample = sample_search_obj.items[0]
    console.log("in select and hover with", sample)
    
    if (sample) {
        // get curve and sample index
        for (let curve_idx=0; curve_idx < pb.data.length; curve_idx++) {
            for (let sample_idx=0; sample_idx < pb.data[curve_idx].text.length; sample_idx++) {
                let b_sample = pb.data[curve_idx].text[sample_idx]
                b_sample = b_sample.substring(7, b_sample.length)
                if (b_sample == sample) {
                    Plotly.Fx.hover(pb, [
                        {curveNumber: curve_idx, pointNumber: sample_idx}
                    ])
                    pa_hover(sample)
                }
            }
        }
    } else {
        // sample de-selected; reset A
        pa_hover(false)
        // reset B
        Plotly.purge(pb)
        init_depth_metrics_plot()
    }
}

function init_depth_metrics_plot() {
    Plotly.react(pb, traces_b, layout_b)

    pb.removeAllListeners('plotly_hover')
    pb.removeAllListeners('plotly_unhover')

    pb.on('plotly_hover',  function(e) {
        var p = e.points[0]
        if (p.text[6] != ":") {
            alert("bad sample name:" + p.text[6])
        }
        var sample = p.text.substring(7, p.text.length)
        pa_hover(sample)
    })

    // remove the markers and fix opacity on unhover
    pb.on('plotly_unhover',  function(e) {
        input[input.length-1].x = [];
        input[input.length-1].y = [];
        input[input.length-1].hovertext = [];
        for(j = 0; j < input.length-1; j++){
            input[j].marker.opacity = 1
        }
        Plotly.react(pa, input, layout_a)
    })
}

// select listeners for plot A
jQuery('#plotay_select, #plotax_select').on('change', function() {
    pa_hover(false)
    var metric = this.value
    set_xy_data_by_group(input, metric, this.id == "plotax_select")

    if(this.id == 'plotay_select') {
        layout_a.yaxis.title = jQuery("#plotay_select option:selected").text();
    } else {
        layout_a.xaxis.title = jQuery("#plotax_select option:selected").text();
    }
    Plotly.react(pa, input, layout_a)
    select_and_hover()
})

// select listeners for plot B
jQuery('#plotbx_select, #plotby_select').on('change', function() {
    var field = this.value
    var extracted = [[], [], []]
    for (i in sample_data) {
        var sex = sample_data[i].sex
        var idx = sex == "male" ? 1 : ((sex == "female")? 2: 0)
        extracted[idx].push(sample_data[i][field])
    }
    if (this.id == 'plotbx_select') {
        layout_b.xaxis.title = jQuery("#plotbx_select option:selected").text()
        traces_b[0].x = extracted[0]
        traces_b[1].x = extracted[1]
        traces_b[2].x = extracted[2]
    } else {
        layout_b.yaxis.title = jQuery("#plotby_select option:selected").text()
        traces_b[0].y = extracted[0]
        traces_b[1].y = extracted[1]
        traces_b[2].y = extracted[2]
    }
    Plotly.react(pb, traces_b, layout_b)

    // check if sample is selected in sample-select dropdown
    select_and_hover()
})

// responsive plots
window.onresize = function() {
    Plotly.Plots.resize('plota')
    Plotly.Plots.resize('plotb')
};

const sample_names = sample_data.map(function(x) {
    return x.sample
})
const male_sample_names = sample_data.filter(function(x) { return x.sex == "male"}).map(function(x) { return x.sample })
const female_sample_names = sample_data.filter(function(x) { return x.sex == "female"}).map(function(x) { return x.sample })
const unknown_sample_names = sample_data.filter(function(x) { return x.sex != "male" && x.sex != "female"}).map(function(x) { return x.sample })

// sample search
const sample_list = sample_data.map(function(x) {
    return { item: x.sample }
})
const sample_search = $('#sample-search').selectize({
    plugins: ['remove_button'],
    valueField: 'item',
    labelField: 'item',
    searchField: 'item',
    options: sample_list,
    placeholder: 'Sample ID',
    mode: 'single',
    closeAfterSelect: false,
})
const sample_search_obj = sample_search[0].selectize

$('#sample-search').on('change', function() {
    select_and_hover()
})
</script>

</body>
</html>
